nomos-core/chain-defs/src/mantle/genesis_tx.rs

```diff
@@ -0,0 +1,390 @@
+use groth16::Fr;
+use poseidon2::Digest;
+use serde::Serialize;
+
+use crate::{
+    crypto::ZkHasher,
+    mantle::{
+        AuthenticatedMantleTx, MantleTx, SignedMantleTx, Transaction, TransactionHasher, TxHash,
+        gas::{Gas, GasConstants, GasCost},
+        ops::{
+            Op, OpProof,
+            channel::{ChannelId, MsgId, inscribe::InscriptionOp},
+        },
+    },
+    proofs::zksig::ZkSignatureProof,
+};
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
+pub struct GenesisTx(SignedMantleTx);
+
+#[derive(Debug, thiserror::Error, Clone, PartialEq, Eq)]
+pub enum Error {
+    #[error("Genesis transaction must have gas price of zero")]
+    InvalidGenesisGasPrice,
+    #[error("Genesis transaction should not have any inputs")]
+    UnepectedInput,
+    #[error("Genesis block cannot contain this op: {0:?}")]
+    UnsupportedGenesisOp(Box<Op>),
+    #[error("Expected exactly one inscription in genesis block")]
+    MissingInscription,
+    #[error("Invalid genesis inscription: {0:?}")]
+    InvalidInscription(Box<Op>),
+}
+
+impl GenesisTx {
+    pub fn from_tx(signed_tx: SignedMantleTx) -> Result<Self, Error> {
+        let mantle_tx = signed_tx.mantle_tx();
+
+        // Genesis transactions must have gas prices of zero
+        if mantle_tx.execution_gas_price != 0 || mantle_tx.storage_gas_price != 0 {
+            return Err(Error::InvalidGenesisGasPrice);
+        }
+
+        // Genesis transactions should not have any inputs
+        if !mantle_tx.ledger_tx.inputs.is_empty() {
+            return Err(Error::UnepectedInput);
+        }
+
+        // Genesis transactions must contain exactly one inscription as the first op
+        // and then may contain other SDP declarations
+        let mut ops = mantle_tx.ops.iter();
+        match ops.next() {
+            Some(Op::ChannelInscribe(op)) => validate_inscription_op(op)?,
+            _ => return Err(Error::MissingInscription),
+        }
+
+        for op in ops {
+            match op {
+                Op::SDPDeclare(_) => {}
+                _ => return Err(Error::UnsupportedGenesisOp(Box::new(op.clone()))),
+            }
+        }
```

Comment:
```rust
if ops.iter().any(|op| !matches!(op, Op::SDPDeclare(_))) {
    return Error()
}
```
Or if we want to report txs, I would check for all the ones that are not allowed instead of just the first one (?)

Reply:
> not sure I get what you mean, could you elaborate?

So if somehow we messed up more than a single tx only the first one will be reported. I would take all of the ones that should not be there to be reported instead.